dali/operators/generic/slice/slice_base.h

```diff
@@ -15,59 +15,118 @@
 #ifndef DALI_OPERATORS_GENERIC_SLICE_SLICE_BASE_H_
 #define DALI_OPERATORS_GENERIC_SLICE_SLICE_BASE_H_
 
+#include <memory>
 #include <tuple>
 #include <utility>
 #include <vector>
+#include "dali/core/any.h"
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
-#include "dali/kernels/scratch.h"
-#include "dali/operators/image/crop/crop_attr.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/kernels/slice/slice_kernel_utils.h"
+#include "dali/operators/generic/slice/out_of_bounds_policy.h"
 #include "dali/pipeline/operator/common.h"
 #include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+#include "dali/util/crop_window.h"
 
 #define SLICE_TYPES (uint8_t, int16_t, uint16_t, int32_t, float, float16)
-#define CMN_NDIMS (3, 4, 5)
+#define SLICE_DIMS (1, 2, 3, 4)
 
 namespace dali {
 
 template <typename Backend>
 class SliceBase : public Operator<Backend> {
  public:
   explicit inline SliceBase(const OpSpec &spec)
-    : Operator<Backend>(spec)
-    , slice_anchors_(batch_size_)
-    , slice_shapes_(batch_size_)
-    , output_type_(spec.GetArgument<DALIDataType>("output_dtype")) {
+      : Operator<Backend>(spec),
+        output_type_(spec.GetArgument<DALIDataType>("output_dtype")),
+        out_of_bounds_policy_(GetOutOfBoundsPolicy(spec)) {
+    if (out_of_bounds_policy_ == OutOfBoundsPolicy::Pad) {
+      fill_values_ = spec.GetRepeatedArgument<float>("fill_values");
+    }
   }
 
- protected:
-  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
-    return false;
-  }
+  template <typename OutputType, int Dims>
+  void FillArgs(std::vector<kernels::SliceArgs<OutputType, Dims>>& slice_args,
+                const workspace_t<Backend> &ws) {
+    this->DataDependentSetup(ws);
```

Comment:
I might be nitpicky, but this will be more general comment to the code structure in the FillArgs function.

The DataDependentSetup does some magic, but in fact it just obtains the cropping information from the arguments (passed in a crop or slice way). Maybe more direct call to Slice/Crop attr here would be clearer, or some other name that highlights that fact.

Shouldn't the fill_values be validated once against the number of channels, preferably by OutOfBoundsAttrs?

This fills a little bit boilerplatey, without the channel business and by adding a function to convert from CropWindow to SliceChannels it would be compacted to basically:

```
this->ProcessCroppingAttrs(ws);
for (samples_idx : samples) {
  auto crop_win_gen = this->GetCropWindowGenerator(i);
  CropWindow win = crop_win_gen(in_shape[i], in_layout);
  ApplySliceBoundsPolicy(out_of_bounds_policy_, in_shape[i], win.anchor, win.shape);
  slice_args[i] = ToSliceArgs(win);
```

Reply:
Basically, your `ToSliceArgs` would do what this `FillArgs` does, except for maybe the validation of fill_values that could happen somewhere else. That and the name change from DataDependentSetup to ProcessCroppingAttrs. Correct?